Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,828 @@
+  using LowestLayer = typename Base::LowestLayer;  // the K=1 class
+
+ private:
+  /*! \brief Number of nodes in this layer.
+   *
+   * For the layer K=1, nNodes will be the number of data points.
+   * For a layer K>1, nNodes will be the number of sublists.
+   */
+  Index_t nNodes = 0;
+
+  /*! \brief Iterator used at construction, runs from 0 to (nNodes-1). */
+  Index_t iNode = 0;
+
+  /*! \brief Indices in the lower layer's indices or data array */
+  std::vector<Index_t> indices;
+
+  /*=== Getters ===*/
+ public:
+  Index_t* GetIndices() { return indices.data(); }
+  const Index_t* GetIndices() const { return indices.data(); }
+
+  /*=== Outputting ===*/
+
+  /*! \brief Write in Python-list style.
+   *
+   * Like this: [[1, 2], [10, 20, 30]]
+   */
+  friend std::ostream& operator<<(std::ostream& output, NdFlattener const& nd) {
+    nd.toPythonString_fromto(output, 0, nd.size());
+    return output;
+  }
+
+ protected:
+  /*! \brief Write to stream in Python-list style, using the data of the
+   * indices array between 'from' (inclusive) and 'to' (exclusive).
+   *
+   * Like this: [[1, 2], [10, 20, 30]]
+   * \param[in] output - Stream
+   * \param[in] from - Beginning of the representation in the indices array.
+   * \param[in] to - Ending of the representation in the indices array.
+   */
+  void toPythonString_fromto(std::ostream& output, Index_t from, Index_t to) const {
+    output << "[";
+    for (Index_t i = from; i < to;) {
+      Base::toPythonString_fromto(output, indices[i], indices[i + 1]);
+      if (++i < to) output << ", ";
+    }
+    output << "]";
+  }
+
+ public:
+  /*! \brief Basic constructor. Afterwards, initialization can be done with initialize_or_refresh.
+   *
+   * Called recursively when a derived class (higher K) is constructed.
+   */
+  NdFlattener() {}
+
+  /*! \brief Constructor which calls initialize_or_refresh.
+   *
+   */
+  template <class... ARGS>
+  NdFlattener(ARGS const&... args) {
+    initialize_or_refresh(args...);
+  }
+
+  /*! \brief Initialize or refresh the NdFlattener.
+   * \details Either a 'recursive function' or 'collective communication'
+   * may be used. When the NdFlattener does not hold data yet, it is
+   * initialized, meaning that the data are collected and the indices arrays
+   * are allocated and filled. Otherwise it is refreshed, meaning that the data are
+   * recollected under the assumption that the indices arrays did not change.
+   */
+  template <class... ARGS>
+  void initialize_or_refresh(ARGS const&... args) {
+    if (initialized()) {
+      refresh(args...);
+    } else {
+      initialize(args...);
+    }
+  }
+
+  /*! \brief Initialization status of the NdFlattener.
+   * \returns true if the NdFlattener has been initialized
+   */
+  bool initialized() { return nNodes > 0; }
+
+ protected:
+  /*! \brief Allocate the indices array after \a nNodes has been determined.
+   */
+  void allocate() {
+    indices.reserve(nNodes + 1);
+    indices[0] = 0;
```

Comment:
I believe this use of vector::reserve is not correct, reserve changes the capacity of the vector but not its size.
Technically indices[0] is out of bounds (not in memory but in "logic").
Whether this is the problem or not IDK.